### PR TITLE
[keras] Fix AUC.result returning 0 for single-class ROC inputs

### LINKED ISSUE #41081: Zero AUROC when validation set is only comprised of positive labels
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Tested on Ubuntu 18.04 and MacOS Catalina 10.15.5
- TensorFlow installed from (source or binary): installed from Conda (tensorflow-gpu) on Ubuntu and from pip on MacOS
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.6 (on both machines)
- CUDA/cuDNN version: 11.0 (on Ubuntu machine)
- GPU model and memory: 4 NVIDIA V100 GPUs

**Describe the current behavior**
The AUROC is zero when there are no negatives in the validation set. This may lead users to believe that the model is not performing correctly, while it is just an error related to a misuse of the AUROC metric.

**Describe the expected behavior**
The AUROC should be NaN when no negatives are present in the validation set. This way the user is aware that there is something wrong with the metric that is being used, with the possibility to add a warning since there is a 0/0 somewhere in the computation of the AUROC.

**Standalone code to reproduce the issue**
```python
from tensorflow.keras.metrics import AUC
metric = AUC()
metric.update_state([1, 1, 1], [1, 0.5, 0.3])
metric.result()
```

I believe that similar issues also happen for other metrics, as uncanny values were displayed also for Recall and Precision, but I have not identified a simple reproducible example.

### PR DESCRIPTION
**Pull request by Ijtihed**
When y_true contains only one class (all-positive or all-negative) ROC-AUC is undefined. AUC.result() returned 0.0 because div_no_nan silently maps 0/0 to 0 making all fp-rate differences zero in the Riemann sum.

Checks true_positives[0] and false_positives[0] at threshold -epsilon (where all examples are predicted positive) and returns NaN if either class is absent. 

Fixes #41081

### CODE REVIEW

**Review by ratgr — CHANGES_REQUESTED** [review]
Thank you for addressing the issue where `AUC.result()` returns `0` instead of `NaN` for single-class inputs! This is a valuable fix for clarity.

  While reviewing the changes, we have a few technical observations regarding edge cases:

  ### Multi-Label Scenarios
  The fix checks if `no_positives` or `no_negatives` and returns `NaN`. However, if `multi_label=True` (and averaging is performed in `result()`), this check seems to apply to the global state or might not handle the case where **only one** of the labels has single-class inputs correctly.
  * If `multi_label=True`, should the metrics for individual labels return `NaN` and be excluded from averaging? Or should the entire result be `NaN`?
  * Currently, if one label is single-class and others are mixed, the single-class one might still contribute `0.0` (due to `div_no_nan` in helpers) or `NaN` which could poison the average depending on how `reduce_mean`/`reduce_sum` handles it. Please verify that `multi_label` support works as intended for this edge case.

  ### Consistency with PR-AUC
  We noticed the discussion regarding leaving `PR-AUC` out of scope. While this is acceptable for now, for consistency, it might be worth verifying if similar issues (e.g., return value for all-negative inputs in PR space) should be addressed or documented similarly in the future.

  ### Custom Thresholds
  The GitHub discussion mentions potential fragility with custom thresholds if using indexes like `[0]`. The sum-based approach (`reduce_sum`) seems more robust. Please ensure this is what is used.

  Actionable Advice: Please verify that the fix behaves correctly (and consistently) when `multi_label=True` and only a subset of labels are single-class.

**Comment by Ijtihed** [comment]
sorry I closed this by accident since I botched a git reset on a shallow clone, rewrote the head and it auto-closed. the branch is fixed now and addresses your points. could you reopen? if github won't let you, i'll open a fresh one and link it. @ratgr

**Comment by ratgr** [comment]
@keerthanakadiri I don't have the privileges to reopen this ticket, can you open it or should they open a fresh one?

**Comment by keerthanakadiri** [comment]
> @keerthanakadiri I don't have the privileges to reopen this ticket, can you open it or should they open a fresh one?

@ratgr Unfortunately, because the ticket was force closed, I don't have the privileges to reopen it on my end either. @Ijtihed Please submit a new ticket. Thank you!
